builtin.c

```diff
@@ -489,35 +492,93 @@ static OP *ck_builtin_funcN(pTHX_ OP *entersubop, GV *namegv, SV *ckobj)
+    { "refaddr",    SHORTVER(5,39), &XS_builtin_func1_scalar, &ck_builtin_func1, OP_REFADDR,    false },
+    { "reftype",    SHORTVER(5,39), &XS_builtin_func1_scalar, &ck_builtin_func1, OP_REFTYPE,    false },
+    { "ceil",       SHORTVER(5,39), &XS_builtin_func1_scalar, &ck_builtin_func1, OP_CEIL,       false },
+    { "floor",      SHORTVER(5,39), &XS_builtin_func1_scalar, &ck_builtin_func1, OP_FLOOR,      false },
+    { "is_tainted", SHORTVER(5,39), &XS_builtin_func1_scalar, &ck_builtin_func1, OP_IS_TAINTED, false },
+    { "trim",       SHORTVER(5,39), &XS_builtin_trim,         &ck_builtin_func1, 0,             false },
+    { "stringify",       NO_BUNDLE, &XS_builtin_func1_scalar, &ck_builtin_func1, OP_STRINGIFY,  true },
+
+    { "created_as_string", NO_BUNDLE, &XS_builtin_created_as_string, &ck_builtin_func1, 0, true },
+    { "created_as_number", NO_BUNDLE, &XS_builtin_created_as_number, &ck_builtin_func1, 0, true },
 
     /* list functions */
-    { "builtin::indexed",          &XS_builtin_indexed,          &ck_builtin_funcN, 0, false },
-    { "builtin::export_lexically", &XS_builtin_export_lexically, NULL,              0, true },
+    { "indexed",          SHORTVER(5,39), &XS_builtin_indexed,          &ck_builtin_funcN, 0, false },
+    { "export_lexically",      NO_BUNDLE, &XS_builtin_export_lexically, NULL,              0, true },
 
-    { NULL, NULL, NULL, 0, false }
+    { NULL, 0, NULL, NULL, 0, false }
 };
 
+static bool S_parse_version(const char *vstr, int *vmajor, int *vminor)
+{
+    /* Parse a string like "5.35" to yield 5 and 35. Ignores an optional
+     * trailing third component e.g. "5.35.7". Returns false on parse errors.
+     */
+
+    size_t len;
+
+    if(sscanf(vstr, "%d.%d%zn", vmajor, vminor, &len) < 2)
+        return FALSE;
+
+    if(*vminor > 255)
+        return FALSE;
+
+    vstr += len;
+
+    if(vstr[0] == '.') {
+        vstr++;
+
+        int _dummy;
+        if(sscanf(vstr, "%d%zn", &_dummy, &len) < 1)
+            return FALSE;
+        if(_dummy > 255)
+            return FALSE;
+
+        vstr += len;
+    }
+
+    if(vstr[0])
+        return FALSE;
+
+    return TRUE;
+}
+
+#define import_sym(sym)  S_import_sym(aTHX_ sym)
+static void S_import_sym(pTHX_ SV *sym)
+{
+    SV *ampname = sv_2mortal(Perl_newSVpvf(aTHX_ "&%" SVf, SVfARG(sym)));
+    SV *fqname = sv_2mortal(Perl_newSVpvf(aTHX_ "builtin::%" SVf, SVfARG(sym)));
+
+    CV *cv = get_cv(SvPV_nolen(fqname), SvUTF8(fqname) ? SVf_UTF8 : 0);
+    if(!cv)
+        Perl_croak(aTHX_ builtin_not_recognised, sym);
+
+    export_lexical(ampname, (SV *)cv);
+}
+
+#define import_builtin_bundle(ver)  S_import_builtin_bundle(aTHX_ ver)
+static void S_import_builtin_bundle(pTHX_ U16 ver)
+{
+    for(int i = 0; builtins[i].name; i++) {
+        if(builtins[i].since_ver <= ver)
+            import_sym(newSVpvn_flags(builtins[i].name, strlen(builtins[i].name), SVs_TEMP));
+    }
+}
```

Comment:
This seems like a lot of busywork to take char *s and put them into new mortal SVs in order to:
* Do some expensive/convoluted concatenations (via `Perl_newSVpvf`, compared with manually building concatenated strings from a pointer to `builtins[i].name` within `S_import_sym`), then:
* Just use the char * of one concatenation (`fqname`) in one call (to `get_cv`)
* Pass the other SV to `S_export_lexical`, which calls `pad_add_name_sv`, which calls `pad_add_name_pvn`, which strips out the char * to use.

Instead:
* Could `S_import_builtin_bundle` pass the char * and strlen as arguments to `S_import_sym`
* If there's no realistic danger of `get_cv` and `export_lexical` croaking (there doesn't seem to be), could `S_import_sym` either manually allocate a buffer to use for the first concatenation and to pass to `get_cv`, then reuse the buffer for the second concatenation and to pass to `export_lexical`, or
* If `get_cv` and `export_lexical` could feasibly croak, `S_import_sym` could create one mortal SV, use it for the `get_cv` call, then reuse it for the second concatenation and `export_lexical` call?
* Could `S_export_lexical` be changed to take a char * and length, then call `pad_add_name_pvn` directly?

PS If your answer is "yes, in theory, but I'm busy", I'm happy to properly look at this and submit a follow-up PR if feasible once this PR is merged.

Reply:
@richardleach I've had a go at implementing this, and while it does reduce the churn of temporary SVs, my current attempt goes for passing around `char *`s instead of `SV *`s, which means they forget the length and UTF8 flag. Not a huge deal currently because of the nature of the symbol names we're working on, but it feels like conceptually a backwards step.

Since all these extra temporary SVs only happen once at import time, I'm tempted to suggest it's not a huge deal - already at compiletime there's so many other temporary SVs and optree fragments floating around that eventually get tidied up before runtime, I don't think an extra few here are going to make a huge difference,

I'm tempted to say merge it as it stands, and if anyone feels particularly enthused they could look into reducing it here, but any sort of work of that nature is likely better spent in other places to bigger impact, honestly.

Reply:
Fair enough, thanks for looking into it.